cylc/flow/subprocpool.py

```diff
@@ -365,16 +365,22 @@ def put_command(
             )
 
     @classmethod
-    def run_command(cls, ctx):
+    def run_command(cls, ctx, timeout: Optional[float] = None):
         """Execute command in ctx and capture its output and exit status.
 
+        Raises subprocess.TimeoutExpired (via subprocess.communicate) if
+        the command gets killed for exceeding a given timeout.
+
         Arguments:
             ctx (cylc.flow.subprocctx.SubProcContext):
                 A context object containing the command to run and its status.
+            timeout:
+                Timeout in seconds, after which to kill the command.
         """
         proc = cls._run_command_init(ctx)
         if proc:
-            ctx.out, ctx.err = (f.decode() for f in proc.communicate())
+            ctx.out, ctx.err = (
+                f.decode() for f in proc.communicate(timeout=timeout))
```

Comment:
The subprocess `timeout` option just raises an exception once the timeout has elapsed. You can catch this exception and retry the `proc.communicate` again because it doesn't actually kill the process. 

> The child process is not killed if the timeout expires, so in order to cleanup properly a well-behaved application should kill the child process and finish communication:
>
> -- https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate

Under normal operation (background rather than foreground) subprocpool implements its own timeout kill:

https://github.com/cylc/cylc-flow/blob/335216488793401ce3f7f90ebf07c11d3187b63c/cylc/flow/subprocpool.py#L280-L291

Reply:
Huh, good spotting. It doesn't stop the scheduler from shutting down anymore, but the subprocess is still running. I just assumed that the `communicate` timeout would kill it. (Isn't that what a "timeout" should do? otherwise it's more of a "timer").

Reply:
Huh, good spotting. It doesn't stop the scheduler from shutting down anymore, but the subprocess is still running. I just assumed that the `communicate` timeout would kill it. (Isn't that what a "timeout" should do? otherwise it's more of a "timer").

Reply:
I've added kill logic to the "foreground" `run_command` function.